### PR TITLE
SONAR-6376 ws plugins/installed and plugins/pending add category field

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java [hunk 1]
```diff
@@ -64,36 +72,42 @@ public void define(WebService.NewController controller) {
...
-    for (PluginInfo pluginMetadata : copyOf(NAME_KEY_PLUGIN_METADATA_COMPARATOR, plugins)) {
-      pluginWSCommons.writePluginMetadata(jsonWriter, pluginMetadata);
+    for (PluginInfo pluginInfo : copyOf(NAME_KEY_PLUGIN_METADATA_COMPARATOR, plugins)) {
+      Plugin plugin = compatiblePluginsByKey.get(pluginInfo.getKey());
+      pluginWSCommons.writePluginInfo(jsonWriter, pluginInfo, categoryOrNull(plugin));
     }
     jsonWriter.endArray();
   }
 
+  private String categoryOrNull(Plugin plugin) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java]: MAJOR Make "categoryOrNull" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstalledAction.java [hunk 2]
```diff
@@ -57,28 +68,41 @@ public void define(WebService.NewController controller) {
...
-      pluginWSCommons.writePluginMetadata(jsonWriter, pluginMetadata);
+    for (PluginInfo pluginInfo : pluginInfoList) {
+      Plugin plugin = compatiblesPluginsByKeys.get(pluginInfo.getKey());
+      pluginWSCommons.writePluginInfo(jsonWriter, pluginInfo, categoryOrNull(plugin));
     }
     jsonWriter.endArray();
   }
+
+  @CheckForNull
+  private String categoryOrNull(Plugin plugin) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstalledAction.java]: MAJOR Make "categoryOrNull" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by sns-seb** [comment]
a bunch of question and comments + waiting for Travis status
back to you @teryk

**Comment by teryk** [comment]
@sns-seb feedback applied in https://github.com/SonarSource/sonarqube/commit/10514c81a7460b2c75c728a5f795449242734a04

**Comment by sns-seb** [comment]
the change that makes the `getUpdateCenter` method return an `Optional` should really be a specific commit for clarity (since it implies many changes in tests which are not those of plugins/installed and plugins/pending)

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstalledAction.java [hunk 3]
```diff
@@ -28,22 +29,28 @@
...
 import static com.google.common.collect.ImmutableSortedSet.copyOf;
 import static org.sonar.server.plugins.ws.PluginWSCommons.NAME_KEY_PLUGIN_METADATA_COMPARATOR;
+import static org.sonar.server.plugins.ws.PluginWSCommons.compatiblePluginsByKey;
 
 /**
  * Implementation of the {@code installed} action for the Plugins WebService.
  */
 public class InstalledAction implements PluginsWsAction {
   private static final String ARRAY_PLUGINS = "plugins";
+  private static final boolean DO_NOT_REFRESH = false;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/InstalledAction.java]: MAJOR Remove this unused "DO_NOT_REFRESH" private field. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)

File: server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PluginWSCommons.java [hunk 4]
```diff
@@ -214,61 +229,55 @@ static String toJSon(PluginUpdate.Status status) {
...
+    return plugin != null ? plugin.getCategory() : null;
+  }
 
-    @Override
-    public String apply(@Nonnull Plugin input) {
-      return input.getName();
-    }
+  private static List<Plugin> compatiblePlugins(UpdateCenterMatrixFactory updateCenterMatrixFactory) {
+    Optional<UpdateCenter> updateCenter = updateCenterMatrixFactory.getUpdateCenter(false);
+    return updateCenter.isPresent() ? updateCenter.get().findAllCompatiblePlugins() : Collections.EMPTY_LIST;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PluginWSCommons.java]: MAJOR Replace "Collections.EMPTY_LIST" by "Collections.emptyList()". [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1596)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by teryk** [comment]
feedback applied in https://github.com/SonarSource/sonarqube/commit/f05f283b3fe37b5854f42f6b9a1559d3cde57a3d

**Comment by teryk** [comment]
I'll make a specific commit concerning the change to return an Optional

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by sns-seb** [comment]
@teryk back to you, some more comments and then LGTM

**Pull request merged**